staging/src/k8s.io/api/resource/v1alpha1/types.go

```diff
@@ -0,0 +1,399 @@
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. If this is not specified, the resource is
+	// not yet allocated.
+	Allocation *AllocationResult `json:"allocation,omitempty" protobuf:"bytes,2,opt,name=allocation"`
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
+	//
+	// There can be at most 32 such reservations. This may get increased in
+	// the future, but not reduced.
+	//
+	// +listType=set
+	ReservedFor []ResourceClaimConsumerReference `json:"reservedFor,omitempty" protobuf:"bytes,3,opt,name=reservedFor"`
+
+	// DeallocationRequested indicates that a ResourceClaim is to be
+	// deallocated.
+	//
+	// The driver then must deallocate this claim and reset the field
+	// together with clearing the Allocation field.
+	//
+	// While DeallocationRequested is set, no new consumers may be added to
+	// ReservedFor.
+	DeallocationRequested bool `json:"deallocationRequested,omitempty" protobuf:"varint,4,opt,name=deallocationRequested"`
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// claim.status.reservedFor.
+const ResourceClaimReservedForMaxSize = 32
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// ResourceHandle contains arbitrary data returned by the driver after a
+	// successful allocation. This is opaque for
+	// Kubernetes. Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The maximum size of this field is 16KiB. This may get
+	// increased in the future, but not reduced.
+	ResourceHandle string `json:"resourceHandle,omitempty" protobuf:"bytes,1,opt,name=resourceHandle"`
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// Setting this field is optional. If null, the resource is available
+	// everywhere.
+	AvailableOnNodes *v1.NodeSelector `json:"availableOnNodes,omitempty" protobuf:"bytes,2,opt,name=availableOnNodes"`
+
+	// Shareable determines whether the resource supports more
```

Comment:
It's gone (that's part of the update that I just pushed).

Reply:
Wait a minute -- the ability to share resources is one of **the** main features that NVIDIA is excited about finally being able to support with DRA. We already have an out-of-tree DRA driver running with this functionality and have been communicating to customers that this is the (long-awaited) path towards finally doing GPU-sharing "the right way" in Kubernetes.

Here is a document explaining how the NVIDIA DRA driver works as well as a demo:
* [Dynamic Resource Allocation (DRA) for GPUs in Kubernetes](https://docs.google.com/document/d/1BNWqgx_SmZDi-va_V31v3DnuVwYnF2EmN7D-O_fB6Oo/edit#heading=h.r2l8al89o2ap)
* [Demo of DRA for GPUs](https://drive.google.com/file/d/1iLg2FEAEilb1dcI27TnB19VYtbcvgKhS/view)

Of course, now that the `ResourceClaimTemplate` has to be a separate object the API will become a bit more verbose than described in this document (because we can no longer embed the template in the pod spec), but at least the functionality will be the same.

Reply:
Let me bring it back by dropping that change, for two reasons:
- The NVIDIA driver uses  `shareable: true` because GPU allocations can be used concurrently on the same node; the "same node" part is handled via node selector.
- We could make it "always shareable" without the field, but having the field allows us to test both scenarios (shared/not shared) in preparation for defining more precise usage restrictions (max per node, max nodes, etc.).